### DISCUSSION TITLE
Where to add nvim-lspconfig on_attach functions?

### DISCUSSION THREAD

**Discussion by paul** [LazyVim/LazyVim#616] [Q&A] [answered]
This is perhaps a gap in the LazyVim documentation, or more likely a gap in my understanding of how this all works together. I'm trying to enable [this snippet from the nvim-lspconfig docs](https://github.com/neovim/nvim-lspconfig/wiki/UI-Customization#show-line-diagnostics-automatically-in-hover-window), and it says to put it into `on_attach()`. With the LazyVim plugin config, I'm not exactly sure how to put that in my plugins config overrides: https://github.com/paul/dotfiles/blob/master/nvim/.config/nvim/lua/plugins/lsp.lua#L25

I've tried a few random stabs in the dark, but none seemed to work. Thanks!

**Comment by dpetka2001** [comment]
I'm not entirely sure if this is correct, but you could try to create an autocmd on `LspAttach` event and then put your code inside the callback function.

**Reply by paul** [reply]
Would you mind elaborating on this a bit? Do I put something in my `lua/config/autocmds.lua` file that looks like one of these? https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/autocmds.lua

I tried copying that code from the lspconfig docs (`vim.api.nvim_create_autocmd("CursorHold", {`), but that didn't do it. Do I need to wrap that in another autocmd for LspAttach?

```lua
vim.api.nvim_create_autocmd("LspAttach", { 
  something something, 
  callback = function() 
    vim.api.nvim_create_autocmd("CursorHold", { rest of that snippet })
  end
})
```

**Reply by dpetka2001** [reply]
I'm not entirely sure if this is the correct way, but that's how it seems to me. So you would have nested autocmd inside the autocmd on the `LspAttach` event. Don't know how good practice that is and am not entirely sure if it would work(I might also be entirely wrong about this since I'm not good with Lua). Go ahead like you posted and check out if it works. The maintainer would be more appropriate to give a definitive answer to this. Maybe you would like to ping him for a correct answer to this?

**Reply by dpetka2001** [reply]
Could you pls try this and check if it works?
```lua
return {
  {
    "neovim/nvim-lspconfig",
    init = function()
      local format = require("lazyvim.plugins.lsp.format")
      format.on_attach = function(client, buf)
        if
          client.config
          and client.config.capabilities
          and client.config.capabilities.documentFormattingProvider == false
        then
          return
        end

        if client.supports_method("textDocument/formatting") then
          vim.api.nvim_create_autocmd("BufWritePre", {
            group = vim.api.nvim_create_augroup("LspFormat." .. buf, {}),
            buffer = buf,
            callback = function()
              if format.autoformat then
                format.format()
              end
            end,
          })
        end

        if client.supports_method("textDocument/formatting") then
          vim.api.nvim_create_autocmd("CursorHold", {
            buffer = buf,
            callback = function()
              local opts = {
                focusable = false,
                close_events = { "BufLeave", "CursorMoved", "InsertEnter", "FocusLost" },
                border = "rounded",
                source = "always",
                prefix = " ",
                scope = "cursor",
              }
              vim.diagnostic.open_float(nil, opts)
            end,
          })
        end
      end
    end,
  },
}
```
I found something similar in @amaanq dotfiles and thought it might work for your case. 

Edit: I updated the initial code because what we're doing overwrites the default `format.on_attach()` of Lazyvim and this disables autoformatting. So I included the initial code of Lazyvim along with the autocmd you want to achieve to keep having autoformat.

**Reply by dpetka2001** [reply]
@folke  Is there a better way to do this? If not, would it be possible to expose a custom empty `on_attach` function (just like `keymaps.on_attach` and `format.on_attach`) in order to avoid meddling with the default code and just overwrite the custom `on_attach` with our customizations and have it required by `util.on_attach` along with `keymaps` and `format`?

**Reply by folke** [reply]
I just posted the solution in a separate answer :)

**Reply by dpetka2001** [reply]
Thanx a lot. Much appreciated 😊

**Comment by AlexanderDKazakov** [comment]
Good topic!
I would appreciate it if we could also discuss how to add system headers for clangd LSP.

My current stage is this:
![image](https://user-images.githubusercontent.com/17707394/233032043-b230aa8f-dc17-4414-a4d3-9d86951faa40.png)

You can see that LSP is working, but at the same time, it is missing system headers, which I have no idea where to add. I tried this: 
![image](https://user-images.githubusercontent.com/17707394/233033084-7282e134-abf8-42da-9ca8-3db488bb5ffe.png)

**Reply by amaanq** [reply]
should be autodetected, you might've messed something up in compile_commands.json but this is not a clangd support thread fyi

**Reply by dpetka2001** [reply]
Do you have `libstd++-12-dev` installed? Soon as I installed that it managed to find the missing headers.

**Comment by folke** [comment, marked as answer]
You can add the code below to your `lua/config/autocmds.lua`:

```lua
vim.api.nvim_create_autocmd("LspAttach", {
  group = vim.api.nvim_create_augroup("lsp_attach_auto_diag", { clear = true }),
  callback = function(args)
    -- the buffer where the lsp attached
    ---@type number
    local buffer = args.buf

    -- create the autocmd to show diagnostics
    vim.api.nvim_create_autocmd("CursorHold", {
      group = vim.api.nvim_create_augroup("_auto_diag", { clear = true }),
      buffer = buffer,
      callback = function()
        local opts = {
          focusable = false,
          close_events = { "BufLeave", "CursorMoved", "InsertEnter", "FocusLost" },
          border = "rounded",
          source = "always",
          prefix = " ",
          scope = "cursor",
        }
        vim.diagnostic.open_float(nil, opts)
      end,
    })
  end,
})
```

Or alternatively, using the **LazyVim** `on_attach` helper:

```lua
require("lazyvim.util").on_attach(function(_, buffer)
  -- create the autocmd to show diagnostics
  vim.api.nvim_create_autocmd("CursorHold", {
    group = vim.api.nvim_create_augroup("_auto_diag", { clear = true }),
    buffer = buffer,
    callback = function()
      local opts = {
        focusable = false,
        close_events = { "BufLeave", "CursorMoved", "InsertEnter", "FocusLost" },
        border = "rounded",
        source = "always",
        prefix = " ",
        scope = "cursor",
      }
      vim.diagnostic.open_float(nil, opts)
    end,
  })
end)
```

**Reply by paul** [reply]
Yup, this works! Thanks!
